File: Cli_Core_JavaFX/src/cli_core_javafx_test/Cmd_Item_Date.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package cli_core_javafx_test;

/**
 *
 * @author mike
 */
public class Cmd_Item_Date extends Cmd_Item_Str {

    @Override
    protected boolean Is_Char_Valid(char c, int pos, int len) {
        switch (pos) {
            case 0:
            case 11:
                return (c == '"');
            case 5:
            case 8:
                return (c == '-');
        }
        return (c >= '0' && c <= '9');
    }

    public Cmd_Item_Date(String text, String help) {
        super(text, help);
        Type = "Date";
    }

    @Override
    public String Debug_Value_Get() {
        return "\"2018-01-02\"";
    }

    @Override
    public Cmd_Item_Valid_Result Parse(String s) {
        Value_Str = s;

        if (s.length() == 0) {
            return Cmd_Item_Valid_Result.CMD_ITEM_EMPTY;
        }

        if (s.length() == 1 && s.charAt(0) == '"') {
            return Cmd_Item_Valid_Result.CMD_ITEM_INCOMPLETE_STR;
        }
        if (s.length() >= 1 && s.charAt(0) != '"') {
            return Cmd_Item_Valid_Result.CMD_ITEM_ERROR;
        }

        for (int pos = 0; pos < s.length(); pos++) {
            if (!Is_Char_Valid(s.charAt(pos), pos, s.length())) {
                return Cmd_Item_Valid_Result.CMD_ITEM_ERROR;
            }
        }

        // Date
        if (s.length() >= 7) {
            int m = s.charAt(6) - '0';
            if (m < 0 || m > 1) {
                return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
            }
            if (s.length() >= 8) {
                String s_mm = s.substring(6, 8);
                int mm = Integer.parseInt(s_mm);
                if (mm == 0 || mm > 12) {
                    return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
                }
            }
        }

        if (s.length() >= 10) {
            int d = s.charAt(9) - '0';
            if (d < 0 || d > 3) {
                return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
            }
            if (s.length() >= 11) {
                String s_dd = s.substring(9, 11);
                int dd = Integer.parseInt(s_dd);
                if (dd == 0 || dd > 31) {
                    return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
                }
            }
        }

        if (s.length() == 2) {
            if (s.charAt(0) == '"' && s.charAt(1) == '"') {
                return Cmd_Item_Valid_Result.CMD_ITEM_ERROR; //Cmd_Item_Valid_Result.CMD_ITEM_OK;
            }
            return Cmd_Item_Valid_Result.CMD_ITEM_INCOMPLETE_STR;
        }

        if (s.charAt(s.length() - 1) != '"') {
            return Cmd_Item_Valid_Result.CMD_ITEM_INCOMPLETE_STR;
        }

        return Cmd_Item_Valid_Result.CMD_ITEM_OK;
    }

}

File: Cli_Core_CS/Cli_Core_CS/Cmd_Item_Date.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Cli_Core_CS
{
    class Cmd_Item_Date : Cmd_Item_Str
    {
        public    Cmd_Item_Date(string text, string help) : base(text, help)
        {
            Type = "Date";
        }

        override public string Debug_Value_Get()
        {
            return "\"2018-01-02\"";
        }

        override protected bool Is_Char_Valid(char c, int pos, int len)
        {
            switch (pos)
            {
                case 0:
                case 11:
                    return (c == '"');
                case 5:
                case 8:
                    return (c == '-' || c == ':');
            }
            return (c >= '0' && c <= '9');
        }

        override public Cmd_Item_Valid_Result Parse(string s)
        {
            Value_Str = s;

            if (s.Length == 0) return Cmd_Item_Valid_Result.CMD_ITEM_EMPTY;

            if (s.Length == 1 && s[0] == '"') return Cmd_Item_Valid_Result.CMD_ITEM_INCOMPLETE_STR;
            if (s.Length >= 1 && s[0] != '"') return Cmd_Item_Valid_Result.CMD_ITEM_ERROR;

            for (int pos = 0; pos < s.Length; pos++)
            {
                if (!Is_Char_Valid(s[pos], pos, s.Length)) return Cmd_Item_Valid_Result.CMD_ITEM_ERROR;
            }

            // Date
            if (s.Length >= 7)
            {
                int m = s[6] - '0';
                if (m < 0 || m > 1) return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
                if (s.Length >= 8)
                {
                    //int mm = atoi(s.c_str() + 6);
                    int mm = 0;
                    string s_mm = s.Substring(6, 2);
                    Int32.TryParse(s_mm, out mm);
                    if (mm == 0 || mm > 12) return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
                }
            }

            if (s.Length >= 10)
            {
                int d = s[9] - '0';
                if (d < 0 || d > 3) return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
                if (s.Length >= 11)
                {
                    //int dd = atoi(s.c_str() + 9);
                    int dd = 0;
                    string s_dd = s.Substring(9, 2);
                    Int32.TryParse(s_dd, out dd);
                    if (dd == 0 || dd > 31) return Cmd_Item_Valid_Result.CMD_ITEM_OUT_OF_RANGE;
                }
            }

            if (s.Length == 2)
            {
                if (s[0] == '"' && s[1] == '"') return Cmd_Item_Valid_Result.CMD_ITEM_ERROR; //Cmd_Item_Valid_Result.CMD_ITEM_OK;
                return Cmd_Item_Valid_Result.CMD_ITEM_INCOMPLETE_STR;
            }

            if (s[s.Length - 1] != '"') return Cmd_Item_Valid_Result.CMD_ITEM_INCOMPLETE_STR;

            return Cmd_Item_Valid_Result.CMD_ITEM_OK;
        }

        override public bool Is_Space_After_Add(string s)
        {
            if (s.Length >= 12)
            {
                Cmd_Item_Valid_Result res = Parse(s);
                if (res == Cmd_Item_Valid_Result.CMD_ITEM_OK)
                    return true;
            }
            return false;
        }
    }
}
